File: core/doubts.py
"""Doubt desk - getting a doubt answered without paying for an API.

Three routes, in order of how little they cost you:

  1. HANDOFF (default, zero setup, zero cost)
     Builds a properly structured prompt with your syllabus context and hands it
     to a free web chat or search you already have access to. Nothing is sent
     from this machine - the link opens in your browser and you stay logged in
     with your own free account.

  2. OLLAMA (zero cost, one-time install, works offline)
     If Ollama is running locally, the tracker talks to it directly and shows the
     answer inline. Detection is automatic. Suggested small models that run on a
     modest laptop are listed in SUGGESTED_MODELS.

  3. HOSTED KEY (optional)
     If you ever get a free-tier key from Groq, Google AI Studio or OpenRouter,
     paste it in settings and answers appear inline. The key is stored in your
     local database and used only for these calls.

Route 1 needs nothing but a browser. Routes 2 and 3 are conveniences.

A doubt is also a signal. Once an answer has been read, the desk asks whether it
helped, and later notices whether the same concept was attempted again and got
right. Both land in the feedback table and feed recommendation ranking: a topic
you keep asking about is a topic the planner should be pushing at you, and a
"did not help" answer should not raise your confidence in that topic.

If a local or hosted model is used it only rephrases and explains. It never
decides what you should study - that stays with recommend.py, which is rules
over your own logs.
"""

import json
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime

from . import content, db, feedback

TIMEOUT = 60

SUGGESTED_MODELS = [
    dict(name="qwen2.5:3b", size="~2 GB", note="Fastest useful option on 8 GB RAM"),
    dict(name="llama3.2:3b", size="~2 GB", note="Good general explanations"),
    dict(
        name="qwen2.5:7b", size="~4.7 GB", note="Noticeably better at maths, needs 16 GB"
    ),
    dict(name="deepseek-r1:7b", size="~4.7 GB", note="Shows reasoning steps, slower"),
]

HANDOFF_TARGETS = [
    dict(
        slug="chatgpt",
        name="ChatGPT",
        url="https://chatgpt.com/?q={q}",
        note="Free tier. Prompt is pre-filled in the URL.",
        prefill=True,
    ),
    dict(
        slug="claude",
        name="Claude",
        url="https://claude.ai/new?q={q}",
        note="Free tier. Prompt is pre-filled in the URL.",
        prefill=True,
    ),
    dict(
        slug="perplexity",
        name="Perplexity",
        url="https://www.perplexity.ai/search?q={q}",
        note="Free, cites sources. Good for checking a formula.",
        prefill=True,
    ),
    dict(
        slug="duckduckgo",
        name="DuckDuckGo AI Chat",
        url="https://duckduckgo.com/?q={q}&ia=chat",
        note="No account needed at all.",
        prefill=True,
    ),
    dict(
        slug="gateoverflow",
        name="GATE Overflow",
        url="https://gateoverflow.in/search?q={q}",
        note="Human answers from GATE aspirants. Search the question text itself.",
        prefill=True,
        short_query=True,
    ),
    dict(
        slug="gemini",
        name="Google Gemini",
        url="https://gemini.google.com/app",
        note="Free tier, but does not accept a pre-filled prompt. Copy the prompt first.",
        prefill=False,
    ),
]

PROMPT_TEMPLATE = """You are helping a GATE CSE aspirant. Be precise and brief.

Subject: {subject}
Topic: {topic}
Their target: {target}
{question_block}
Their doubt:
{body}

Answer in this structure:
1. Direct answer in two sentences.
2. The reasoning, step by step, showing every formula used.
3. The single mistake most students make here.
4. One GATE-style follow-up question to test whether the idea landed, with the answer hidden at the very end.

Use plain text. No markdown tables. Keep it under 400 words."""


def build_prompt(conn, body, subject_slug="", topic_slug="", question_id=""):
    subject_name = subject_slug or "not specified"
    topic_name = topic_slug or "not specified"

    if subject_slug:
        row = conn.execute(
            "SELECT name FROM subjects WHERE slug = ?", (subject_slug,)
        ).fetchone()
        if row:
            subject_name = row["name"]
    if topic_slug:
        row = conn.execute(
            "SELECT name FROM topics WHERE slug = ? LIMIT 1", (topic_slug,)
        ).fetchone()
        if row:
            topic_name = row["name"]

    settings = db.get_settings(conn)
    targets = {t["slug"]: t for t in content.targets().get("targets", [])}
    primary = targets.get(settings.get("primary_target", ""), {})
    target_line = primary.get("name", settings.get("exam_name", "GATE CSE"))

    question_block = ""
    if question_id:
        q = content.question_bank().get(question_id)
        if q:
            lines = ["", "The question they are stuck on:", q.get("text", "")]
            for i, opt in enumerate(q.get("options", [])):
                lines.append("  (%s) %s" % (chr(65 + i), opt))
            lines.append("")
            question_block = "\n".join(lines)

    return PROMPT_TEMPLATE.format(
        subject=subject_name,
        topic=topic_name,
        target=target_line,
        question_block=question_block,
        body=body.strip() or "(no detail given)",
    )


def handoff_links(prompt, question_text=""):
    out = []
    for t in HANDOFF_TARGETS:
        if not t.get("prefill"):
            out.append(dict(t, ready_url=t["url"]))
            continue
        payload = question_text if t.get("short_query") and question_text else prompt
        encoded = urllib.parse.quote(payload[:1800])
        out.append(dict(t, ready_url=t["url"].format(q=encoded)))
    return out


# ---------------------------------------------------------------------------
# local Ollama
# ---------------------------------------------------------------------------
def _get_json(url, timeout=4):
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.load(resp)


def probe_ollama(base_url):
    base = (base_url or "http://127.0.0.1:11434").rstrip("/")
    try:
        data = _get_json(base + "/api/tags")
    except (urllib.error.URLError, OSError, ValueError, TimeoutError) as exc:
        return dict(
            available=False,
            base_url=base,
            reason=str(exc),
            install_hint="Install from ollama.com, then run: ollama pull qwen2.5:3b",
            suggested=SUGGESTED_MODELS,
        )
    models = [m.get("name", "") for m in (data.get("models") or [])]
    return dict(
        available=True,
        base_url=base,
        models=models,
        suggested=SUGGESTED_MODELS,
        note=(
            "Running locally. No data leaves this machine and there is no cost."
            if models
            else "Ollama is running but has no models. Run: ollama pull qwen2.5:3b"
        ),
    )


def ask_ollama(base_url, model, prompt):
    base = (base_url or "http://127.0.0.1:11434").rstrip("/")
    payload = json.dumps(dict(model=model, prompt=prompt, stream=False)).encode("utf-8")
    req = urllib.request.Request(
        base + "/api/generate",
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=TIMEOUT * 3) as resp:
        data = json.load(resp)
    return (data.get("response") or "").strip()


# ---------------------------------------------------------------------------
# optional hosted providers (free tiers)
# ---------------------------------------------------------------------------
HOSTED = {
    "groq": dict(
        label="Groq",
        signup="https://console.groq.com/keys",
        default_model="llama-3.3-70b-versatile",
        note="Generous free tier, very fast.",
    ),
    "gemini": dict(
        label="Google AI Studio",
        signup="https://aistudio.google.com/apikey",
        default_model="gemini-2.0-flash",
        note="Free tier with a daily request cap.",
    ),
    "openrouter": dict(
        label="OpenRouter",
        signup="https://openrouter.ai/keys",
        default_model="meta-llama/llama-3.3-70b-instruct:free",
        note="Aggregator. Models ending in :free cost nothing.",
    ),
}


def _post_json(url, payload, headers, timeout=TIMEOUT):
    req = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        headers=dict({"Content-Type": "application/json"}, **headers),
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.load(resp)


def ask_hosted(provider, key, model, prompt):
    if provider not in HOSTED:
        raise ValueError("Unknown provider: %s" % provider)
    if not key:
        raise ValueError("No API key saved for %s." % HOSTED[provider]["label"])
    model = model or HOSTED[provider]["default_model"]

    if provider == "gemini":
        url = (
            "https://generativelanguage.googleapis.com/v1beta/models/"
            "%s:generateContent?key=%s" % (model, urllib.parse.quote(key))
        )
        data = _post_json(url, dict(contents=[dict(parts=[dict(text=prompt)])]), {})
        cands = data.get("candidates") or []
        if not cands:
            raise ValueError("The model returned no candidates.")
        parts = (cands[0].get("content") or {}).get("parts") or []
        return "".join(p.get("text", "") for p in parts).strip()

    url = {
        "groq": "https://api.groq.com/openai/v1/chat/completions",
        "openrouter": "https://openrouter.ai/api/v1/chat/completions",
    }[provider]
    data = _post_json(
        url,
        dict(model=model, messages=[dict(role="user", content=prompt)], max_tokens=1200),
        {"Authorization": "Bearer %s" % key},
    )
    choices = data.get("choices") or []
    if not choices:
        raise ValueError("The model returned no choices.")
    return (choices[0].get("message") or {}).get("content", "").strip()


# ---------------------------------------------------------------------------
# orchestration
# ---------------------------------------------------------------------------
def status(conn):
    settings = db.get_settings(conn)
    provider = settings.get("ai_provider", "none")
    ollama = probe_ollama(settings.get("ollama_url"))
    return dict(
        provider=provider,
        model=settings.get("ai_model", ""),
        has_key=bool(settings.get("ai_key")),
        ollama=ollama,
        hosted={
            k: dict(v, configured=(provider == k and bool(settings.get("ai_key"))))
            for k, v in HOSTED.items()
        },
        inline_available=ollama.get("available")
        or (provider in HOSTED and bool(settings.get("ai_key"))),
    )


def answer_inline(conn, prompt):
    """Try whatever is configured. Returns (answer, provider_used)."""
    settings = db.get_settings(conn)
    provider = settings.get("ai_provider", "none")

    if provider in HOSTED and settings.get("ai_key"):
        return (
            ask_hosted(provider, settings["ai_key"], settings.get("ai_model"), prompt),
            provider,
        )

    if provider in ("ollama", "none", "", None):
        probe = probe_ollama(settings.get("ollama_url"))
        if probe.get("available") and probe.get("models"):
            model = settings.get("ai_model") or probe["models"][0]
            return ask_ollama(settings.get("ollama_url"), model, prompt), "ollama"
        if provider == "ollama":
            raise ValueError(
                "Ollama is not reachable at %s. Start it, or use the handoff links below "
                "which need no setup." % probe.get("base_url")
            )

    raise ValueError(
        "No inline answer source is configured. Use the handoff links below - they are "
        "free and need no key - or install Ollama for inline answers."
    )


def save(
    conn, body, subject_slug, topic_slug, question_id, prompt, provider="", answer=""
):
    now = datetime.now().isoformat(timespec="seconds")
    with conn:
        cur = conn.execute(
            "INSERT INTO doubts (subject_slug, topic_slug, question_id, body, prompt,"
            " provider, answer, created_at) VALUES (?,?,?,?,?,?,?,?)",
            (
                subject_slug or "",
                topic_slug or "",
                question_id or "",
                body,
                prompt,
                provider,
                answer,
                now,
            ),
        )
        return cur.lastrowid


def rate(conn, doubt_id, helped, note=""):
    """Record whether the answer actually helped, and feed it to the ranker."""
    row = conn.execute("SELECT * FROM doubts WHERE id = ?", (int(doubt_id),)).fetchone()
    if not row:
        raise LookupError("That doubt is not in the log.")
    now = datetime.now().isoformat(timespec="seconds")
    with conn:
        conn.execute(
            "UPDATE doubts SET helped = ?, resolved = ?, resolved_at = ? WHERE id = ?",
            (
                1 if helped else 0,
                1 if helped else 0,
                now if helped else None,
                int(doubt_id),
            ),
        )
    feedback.record(
        conn,
        "doubt",
        "helped" if helped else "did_not_help",
        target_id=str(doubt_id),
        target_slug="doubt",
        subject_slug=row["subject_slug"],
        topic_slug=row["topic_slug"],
        reason=note[:280],
    )
    return dict(id=int(doubt_id), helped=bool(helped))


def check_retries(conn):
    """Mark a doubt as revisited once the topic has been attempted again.

    This is the delayed half of the signal: "did the explanation stick" is a
    better measure than "did it feel good at the time".
    """
    now = datetime.now().isoformat(timespec="seconds")
    rows = conn.execute(
        "SELECT id, topic_slug, created_at FROM doubts"
        " WHERE retried_at IS NULL AND topic_slug != ''"
    ).fetchall()
    touched = 0
    with conn:
        for r in rows:
            hit = conn.execute(
                "SELECT 1 FROM attempts WHERE topic_slug = ? AND created_at > ?"
                " AND correct = 1 LIMIT 1",
                (r["topic_slug"], r["created_at"]),
            ).fetchone()
            if hit:
                conn.execute(
                    "UPDATE doubts SET retried_at = ? WHERE id = ?", (now, r["id"])
                )
                touched += 1
    return touched


def topic_pressure(conn, days=60):
    """Topics you keep asking about. A useful weak-topic signal on its own."""
    from datetime import timedelta

    since = (datetime.now() - timedelta(days=days)).isoformat(timespec="seconds")
    rows = conn.execute(
        "SELECT subject_slug, topic_slug, COUNT(*) n,"
        " SUM(COALESCE(helped, 0)) helped, SUM(retried_at IS NOT NULL) retried"
        " FROM doubts WHERE created_at >= ? AND topic_slug != ''"
        " GROUP BY subject_slug, topic_slug HAVING n >= 2 ORDER BY n DESC LIMIT 8",
        (since,),
    ).fetchall()
    return [
        dict(
            subject=r["subject_slug"],
            topic=r["topic_slug"],
            asked=r["n"],
            helped=r["helped"],
            retried=r["retried"],
            unresolved=r["n"] - r["helped"],
        )
        for r in rows
    ]


def history(conn, limit=30):
    rows = conn.execute(
        "SELECT * FROM doubts ORDER BY id DESC LIMIT ?", (limit,)
    ).fetchall()
    out = []
    for r in rows:
        item = dict(r)
        item["helped"] = None if r["helped"] is None else bool(r["helped"])
        item["retried"] = bool(r["retried_at"])
        out.append(item)
    return out

File: static/js/doubts.js
/* Doubt desk: prompt composer, handoff links, optional inline answer. */
GG.doubts = (function () {
  'use strict';
  const { $, $$, el, ago } = GG;

  let lastPrompt = '';
  let ctx = { subject_slug: '', topic_slug: '', question_id: '' };

  function tab(st) {
    const sSel = $('#dSubject');
    if (sSel.dataset.built !== String(st.subjects.length)) {
      sSel.innerHTML = '<option value="">Not specified</option>';
      st.subjects.forEach(s => sSel.appendChild(
        el('option', { value: s.slug, text: s.name })));
      sSel.dataset.built = String(st.subjects.length);
    }
    sSel.onchange = () => fillTopics(st);
    fillTopics(st);

    $('#dBuild').onclick = () => build(false);
    $('#dAsk').onclick = () => build(true);

    refreshStatus();
    history(st);
  }

  function fillTopics(st) {
    const tSel = $('#dTopic');
    const slug = $('#dSubject').value;
    tSel.innerHTML = '<option value="">Not specified</option>';
    const subj = st.subjects.find(s => s.slug === slug);
    if (subj) subj.topics.forEach(t => tSel.appendChild(
      el('option', { value: t.slug, text: t.name })));
  }

  async function refreshStatus() {
    let s;
    try { s = await GG.api('/doubts/status'); } catch (e) { return; }
    const line = $('#dStatusLine');
    if (s.inline_available) {
      line.innerHTML = 'Inline answers ready via <b>' +
        (s.ollama.available ? 'local Ollama' : GG.esc(s.provider)) + '</b>';
    } else {
      line.textContent = 'No inline source set up. Handoff links below work with no setup.';
    }
    providers(s);
    return s;
  }

  function providers(s) {
    const host = $('#dProviders');
    host.innerHTML = '';

    /* local */
    const ol = s.ollama;
    host.appendChild(el('div', { style: 'margin-bottom:18px' }, [
      el('div', { class: 'row-end', style: 'margin:0 0 10px' }, [
        el('b', { style: 'font-size:13px', text: 'Route 1 - local model (free, offline)' }),
        el('span', { class: 'state-key ' + (ol.available ? 'live' : 'off'),
                     text: ol.available ? 'DETECTED' : 'NOT RUNNING' }),
      ]),
      el('p', { class: 'dim small', style: 'margin:0 0 10px',
        text: ol.available
          ? (ol.models && ol.models.length
              ? 'Ollama is up at ' + ol.base_url + ' with: ' + ol.models.join(', ')
              : 'Ollama is up but has no model pulled yet.')
          : 'Install Ollama from ollama.com, then run "ollama pull qwen2.5:3b". ' +
            'It runs on your own machine, costs nothing and works without internet.' }),
      el('div', { class: 'row-gap' }, [
        el('button', { class: 'btn btn-sm', text: 'Recheck', onclick: refreshStatus }),
        el('button', { class: 'btn btn-sm', text: 'Configure', onclick: () => aiModal(s) }),
      ]),
      el('div', { class: 'prov-grid', style: 'margin-top:12px' },
        (ol.suggested || []).map(m => el('div', { class: 'prov' }, [
          el('span', { class: 'prov-name mono', text: m.name }),
          el('span', { class: 'prov-note', text: m.note }),
          el('span', { class: 'state-key', text: m.size }),
        ]))),
    ]));

    /* hosted */
    host.appendChild(el('div', { style: 'margin-bottom:18px' }, [
      el('div', { class: 'row-end', style: 'margin:0 0 10px' }, [
        el('b', { style: 'font-size:13px', text: 'Route 2 - hosted free tier (optional key)' }),
        el('span', { class: 'state-key ' + (s.has_key ? 'live' : 'off'),
                     text: s.has_key ? 'KEY SAVED' : 'NO KEY' }),
      ]),
      el('div', { class: 'prov-grid' },
        Object.keys(s.hosted).map(k => {
          const h = s.hosted[k];
          return el('div', { class: 'prov' }, [
            el('span', { class: 'prov-name', text: h.label }),
            el('span', { class: 'prov-note', text: h.note }),
            el('div', { class: 'prov-foot' }, [
              el('a', { class: 'link', href: h.signup, target: '_blank',
                        rel: 'noopener', text: 'Get a free key' }),
              h.configured ? el('span', { class: 'state-key live', text: 'ACTIVE' }) : null,
            ]),
          ]);
        })),
    ]));

    /* handoff */
    host.appendChild(el('div', {}, [
      el('div', { class: 'row-end', style: 'margin:0 0 10px' }, [
        el('b', { style: 'font-size:13px', text: 'Route 3 - handoff (always available, no setup)' }),
        el('span', { class: 'state-key live', text: 'READY' }),
      ]),
      el('p', { class: 'dim small', style: 'margin:0',
        text: 'Build a prompt above and these open with it pre-filled, in whatever free ' +
              'account you already have. Nothing leaves this machine until you click.' }),
    ]));
  }

  async function build(inline) {
    const body = $('#dBody').value.trim();
    if (!body) { GG.toast('Write the doubt first', 'Even one line is enough.', 'bad'); return; }
    ctx.subject_slug = $('#dSubject').value;
    ctx.topic_slug = $('#dTopic').value;

    const btn = inline ? $('#dAsk') : $('#dBuild');
    const label = btn.textContent;
    btn.disabled = true;
    btn.innerHTML = '<span class="spinner"></span> working';

    try {
      const built = await GG.api('/doubts/prompt', {
        body: Object.assign({ body }, ctx),
      });
      lastPrompt = built.prompt;
      render(built);
      if (inline) {
        const out = await GG.api('/doubts/ask', {
          body: Object.assign({ body, prompt: built.prompt }, ctx),
        });
        showAnswer(out);
        GG.toast('Answered', 'via ' + out.provider);
      }
    } catch (e) {
      GG.toast(inline ? 'Inline answer failed' : 'Could not build prompt', e.message, 'bad');
    } finally {
      btn.disabled = false;
      btn.textContent = label;
    }
  }

  function render(built) {
    const host = $('#dResult');
    host.innerHTML = '';
    const card = el('div', { class: 'card', style: 'margin-top:16px' });
    card.appendChild(el('div', { class: 'row-end', style: 'margin:0 0 12px' }, [
      el('b', { style: 'font-size:13px', text: 'Composed prompt' }),
      el('div', { class: 'row-gap' }, [
        el('button', {
          class: 'btn btn-sm', text: 'Copy prompt',
          onclick: async ev => {
            try {
              await navigator.clipboard.writeText(built.prompt);
              ev.currentTarget.textContent = 'Copied';
              setTimeout(() => (ev.currentTarget.textContent = 'Copy prompt'), 1600);
            } catch (e) { GG.toast('Copy blocked', 'Select the text manually.', 'bad'); }
          },
        }),
      ]),
    ]));
    card.appendChild(el('div', { class: 'prompt-box', text: built.prompt }));
    card.appendChild(el('p', { class: 'dim small', style: 'margin:14px 0 8px',
      text: 'Open it somewhere free:' }));
    card.appendChild(el('div', { class: 'prov-grid' },
      built.links.map(l => el('div', { class: 'prov' }, [
        el('span', { class: 'prov-name', text: l.name }),
        el('span', { class: 'prov-note', text: l.note }),
        el('a', { class: 'btn btn-sm', href: l.ready_url, target: '_blank', rel: 'noopener',
                  text: l.prefill ? 'Open with prompt' : 'Open (paste it)' }),
      ]))));
    host.appendChild(card);
    host.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
  }

  function showAnswer(out) {
    const host = $('#dResult');
    host.appendChild(el('div', { class: 'card', style: 'margin-top:16px' }, [
      el('div', { class: 'row-end', style: 'margin:0 0 12px' }, [
        el('b', { style: 'font-size:13px', text: 'Answer' }),
        el('span', { class: 'state-key live', text: out.provider.toUpperCase() }),
      ]),
      el('div', { class: 'answer-box', text: out.answer }),
      el('p', { class: 'dim small', style: 'margin:12px 0 0',
        text: 'A language model wrote this. Check it against a standard text before you ' +
              'trust it on exam day.' }),
    ]));
    if (out.items) historyRows(out.items);
  }

  function history(st) {
    GG.api('/doubts').then(d => historyRows(d.items)).catch(() => {});
  }

  function historyRows(items) {
    const host = $('#dHistory');
    host.innerHTML = '';
    if (!items || !items.length) {
      host.appendChild(GG.render.emptyBox('No doubts logged',
        'Anything you ask gets kept here with its answer.'));
      return;
    }
    items.forEach(d => {
      host.appendChild(el('div', { class: 'row' }, [
        el('span', { class: 'mono dim', text: d.answer ? '[+]' : '[?]' }),
        el('div', { class: 'row-body' }, [
          el('div', { class: 'row-l1' }, [
            el('span', { class: 'row-name', text: (d.body || '').slice(0, 80) }),
            d.provider ? el('span', { class: 'chip-kind', text: d.provider }) : null,
          ]),
          el('div', { class: 'row-sub',
                      text: [d.subject_slug, d.topic_slug].filter(Boolean).join(' / ') || 'general' }),
        ]),
        el('span', { class: 'row-num', text: d.answer ? 'answered' : 'prompt only' }),
        el('span', { class: 'row-day' }, [
          document.createTextNode(ago((d.created_at || '').slice(0, 10))),
          el('button', { class: 'x', text: '\u00d7', title: 'Delete',
            onclick: async () => {
              try { const r = await GG.api('/doubts/' + d.id, { method: 'DELETE' });
                    historyRows(r.items); } catch (e) {}
            } }),
        ]),
      ]));
    });
  }

  /* jump here from a quiz result */
  function prefill(q) {
    location.hash = '#/doubts';
    setTimeout(() => {
      ctx = { subject_slug: q.subject, topic_slug: q.topic, question_id: q.id };
      const sSel = $('#dSubject');
      sSel.value = q.subject;
      sSel.dispatchEvent(new Event('change'));
      setTimeout(() => { $('#dTopic').value = q.topic; }, 30);
      $('#dBody').value = 'I got this one wrong. Walk me through it from the start.';
      $('#dBody').focus();
    }, 120);
  }

  function aiModal(s) {
    const body = $('#aiBody');
    body.innerHTML = '';
    const providerSel = el('select', { id: 'aiProvider' }, [
      el('option', { value: 'none', text: 'Auto (local Ollama if running, else handoff)' }),
      el('option', { value: 'ollama', text: 'Ollama - local, free' }),
      el('option', { value: 'groq', text: 'Groq - hosted free tier' }),
      el('option', { value: 'gemini', text: 'Google AI Studio - hosted free tier' }),
      el('option', { value: 'openrouter', text: 'OpenRouter - hosted, :free models' }),
    ]);
    providerSel.value = s.provider || 'none';
    body.appendChild(el('label', { class: 'field' }, [
      el('span', { text: 'Source' }), providerSel]));
    body.appendChild(el('label', { class: 'field' }, [
      el('span', { text: 'Model', html: '' }),
      el('input', { type: 'text', id: 'aiModel', value: s.model || '',
                    placeholder: 'e.g. qwen2.5:3b or llama-3.3-70b-versatile' })]));
    body.appendChild(el('label', { class: 'field' }, [
      el('span', { text: 'API key (only for hosted sources)' }),
      el('input', { type: 'password', id: 'aiKey', placeholder: s.has_key ? 'saved - leave blank to keep' : '' })]));
    body.appendChild(el('label', { class: 'field' }, [
      el('span', { text: 'Ollama address' }),
      el('input', { type: 'text', id: 'aiOllama', value: s.ollama.base_url || 'http://127.0.0.1:11434' })]));
    body.appendChild(el('p', { class: 'dim small', style: 'margin:0',
      text: 'The key is stored in your local database and stripped from any backup you export. ' +
            'You never need one - the handoff links work without it.' }));

    $('#saveAi').onclick = async () => {
      const payload = {
        ai_provider: $('#aiProvider').value,
        ai_model: $('#aiModel').value.trim(),
        ollama_url: $('#aiOllama').value.trim(),
      };
      const k = $('#aiKey').value.trim();
      if (k) payload.ai_key = k;
      try {
        await GG.api('/settings/ai', { body: payload });
        GG.modal('#modalAi', false);
        GG.toast('Saved', 'Answer source updated.');
        refreshStatus();
      } catch (e) { GG.toast('Could not save', e.message, 'bad'); }
    };
    GG.modal('#modalAi', true);
  }

  return { tab, prefill, refreshStatus };
})();

/* ==========================================================================
   v3 additions - the "did it actually help?" loop.

   A doubt is only resolved when the topic sticks, so there are two signals here:
   an immediate helped / did-not-help button, and a delayed one the engine sets
   by itself when you later answer that topic correctly.
   ========================================================================== */
(function () {
  const { $, el, api, toast, empty, growBars } = GG;
  const base = GG.doubts;

  function pressure(st) {
    const host = $('#dPressure');
    if (!host) return;
    host.innerHTML = '';
    const rows = (st.doubt_pressure || (st.doubts || {}).pressure || []);
    if (!rows.length) {
      host.appendChild(el('p', { class: 'dim small' }, [
        'Asking about the same topic twice is a stronger weak-topic signal than a single ' +
        'wrong answer. Once that happens, the topic shows up here and gets pushed up in ' +
        'your plan.',
      ]));
      return;
    }
    rows.forEach(r => {
      host.appendChild(el('div', { class: 'bar-row' }, [
        el('span', { text: r.name || r.topic }),
        el('div', { class: 'bar-track' }, [
          el('i', { dataset: { w: String(Math.min(100, r.count * 25)) } })]),
        el('span', { class: 'bar-val', text: r.count + 'x' }),
      ]));
    });
  }

  async function rate(id, helped, node) {
    try {
      await api('/doubts/rate', { method: 'POST', body: { id: id, helped: helped } });
      toast(helped ? 'Good' : 'Noted',
        helped ? 'Marked as resolved.'
               : 'That topic has been pushed up your weak list.');
      if (node) {
        node.innerHTML = '';
        node.appendChild(el('span', { class: 'state ' + (helped ? 'ok' : 'weak') }, [
          el('i', {}, [helped ? '#' : '~']), helped ? 'helped' : 'did not help',
        ]));
      }
    } catch (e) { toast('Could not save', e.message, 'bad'); }
  }

  /* Attach rating buttons to every logged doubt that has no verdict yet. */
  function decorate() {
    GG.$$('#dHistory .row').forEach(row => {
      if (row.dataset.rated) return;
      const id = row.dataset.doubtId;
      if (!id) return;
      row.dataset.rated = '1';
      const box = el('div', { class: 'row-gap' });
      box.appendChild(el('button', { class: 'btn btn-sm', onclick: () => rate(id, true, box) },
        ['Helped']));
      box.appendChild(el('button', { class: 'btn btn-sm', onclick: () => rate(id, false, box) },
        ['Still stuck']));
      row.appendChild(box);
    });
  }

  const originalTab = base.tab;
  base.tab = function (st) {
    originalTab(st);
    pressure(st);
    decorate();
    growBars();
  };
  base.rate = rate;
})();
